server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -216,6 +225,46 @@ private long removeBlipCandidates(GarbageCollectionEnvironment gce,
     return blipCount;
   }
 
+  @VisibleForTesting
+  /**
+   * Double check no tables were missed during GC
+   */
+  protected void ensureAllTablesChecked(Set<TableId> tableIdsBefore, Set<TableId> tableIdsSeen,
+      Set<TableId> tableIdsAfter) {
+
+    // if a table was added or deleted during this run, it is acceptable to not
+    // have seen those tables ids when scanning the metadata table. So get the intersection
+    final Set<TableId> tableIdsMustHaveSeen = new HashSet<>(tableIdsBefore);
+    tableIdsMustHaveSeen.retainAll(tableIdsAfter);
+
+    if (tableIdsMustHaveSeen.isEmpty() && !tableIdsSeen.isEmpty()) {
+      throw new RuntimeException("Garbage collection will not proceed because "
+          + "table ids were seen in the metadata table and none were seen Zookeeper. "
+          + "This can have two causes. First, total number of tables going to/from "
+          + "zero during a GC cycle will cause this. Second, it could be caused by "
+          + "corruption of the metadata table and/or Zookeeper. Only the second cause "
+          + "is problematic, but there is no way to distinguish between the two causes "
+          + "so this GC cycle will not proceed. The first cause should be transient "
+          + "and one would not expect to see this message repeated in subsequent GC cycles.");
+    }
+
+    // From that intersection, remove all the table ids that were seen.
+    tableIdsMustHaveSeen.removeAll(tableIdsSeen);
+
+    // If anything is left then we missed a table and may not have removed rfiles references
+    // from the candidates list that are acutally still in use, which would
+    // result in the rfiles being deleted in the next step of the GC process
+    if (!tableIdsMustHaveSeen.isEmpty()) {
+      log.error("TableIDs before: " + tableIdsBefore);
+      log.error("TableIDs after : " + tableIdsAfter);
+      log.error("TableIDs seen  : " + tableIdsSeen);
+      log.error("TableIDs that should have been seen but were not: " + tableIdsMustHaveSeen);
+      // maybe a scan failed?
+      throw new RuntimeException(
+          "Saw table IDs in ZK that were not in metadata table:  " + tableIdsMustHaveSeen);
+    }
```

Comment:
Please don't introduce new code with generic RTE being thrown. A more specific RTE is always available and likely more appropriate.

Also, this situation is likely to be common if creating and deleting a lot of tables. The canonical determination of whether a table exists or not is that it has an entry in ZK... this is created before metadata entries, and is the last thing removed when a table is deleted. So, it's quite normal and routine for a table ID not to exist in the metadata table if it was just created or deleted. This is not necessarily a problem and shouldn't throw an exception... or we're going to see this a lot. We can probably start over gathering candidates... but we shouldn't just crash.

Reply:
> The canonical determination of whether a table exists or not is that it has an entry in ZK... this is created before metadata entries, and is the last thing removed when a table is deleted.

Good catch, we need to consider table states to avoid this problem.  I mentioned table states in #1377, but its been so long I had completely forgotten about that edge case and I did not reread the issue until now.

When a table is created the following is done.

 1. table is put in ZK w/ TableState.NEW
 2. metadata table is populated
 3. tables state is set to TableState.ONLINE or TableState.OFFLINE

When a table is deleted the following is done.

 1. Table state is set to TableState.DELETING
 2. entries are removed from metadata table
 3. entries are removed from ZK

So from the perspective of GC, if we see a table with a state of TableState.ONLINE or TableState.OFFLINE before and after scanning the metadata table, then it must be seen in the metadata table unless there is a problem.

We need to get a `Map<TableId,TableState>` to properly do this check.

Reply:
Addressed the generic RTE in 1f8d37305789fa1540158e32b54da86a7af3ae87. Implemented the table state checking in the same commit.